src/aero_kernel/src/socket/unix.rs

```diff
@@ -17,16 +17,64 @@
  * along with Aero. If not, see <https://www.gnu.org/licenses/>.
  */
 
-use alloc::sync::Arc;
+use aero_syscall::SocketAddr;
+use alloc::{
+    string::String,
+    sync::{Arc, Weak},
+};
 
-use crate::fs::inode::INodeInterface;
+use crate::fs;
+use crate::fs::{
+    inode::{DirEntry, FileType, INodeInterface, Metadata},
+    FileSystemError, Path, Result,
+};
 
-pub struct UnixSocket;
+pub struct UnixSocket {
+    weak: Weak<UnixSocket>,
+}
 
 impl UnixSocket {
     pub fn new() -> Arc<UnixSocket> {
-        Arc::new(UnixSocket)
+        Arc::new_cyclic(|weak| UnixSocket { weak: weak.clone() })
     }
 }
 
-impl INodeInterface for UnixSocket {}
+impl INodeInterface for UnixSocket {
+    fn metadata(&self) -> Result<Metadata> {
+        Ok(Metadata {
+            id: 0, // FIXME: What should this be?
+            file_type: FileType::Socket,
+            size: 0,
+            children_len: 0,
+        })
+    }
+
+    fn bind(&self, address: &SocketAddr, _length: usize) -> Result<()> {
+        if let SocketAddr::Unix(address) = address {
+            let path_len = address
+                .path
+                .iter()
+                .position(|&b| b == 0)
+                .unwrap_or(address.path.len());
+
+            let path_str = unsafe { core::str::from_utf8_unchecked(&address.path[0..path_len]) };
```

Comment:
We probably dont want to parse the path here. Instead parse it in the syscall and make a new `SocketAddr` enum for the kernel itself. Since we would want to return `EPERM` and stuff if the path's address is larger then the max userland address.

Reply:
We don't know what kind of socket we're dealing with in the `sys_bind`, all we have is an address and a file descriptor. We let the `bind` method on the socket type decide whether the address is valid or not. This should not be done in the syscall itself due to differences between the way each kind of socket treats the address, and depending on that we could end up returning a misleading/bad error code back to the caller.